Enable HMR without script tag in source html

faergeek opened the pull request:
# ↪️ Pull Request

This is a brute-force solution for https://github.com/parcel-bundler/parcel/issues/943. As explained, HMR code is injected into js bundles, and if there's no js bundle, HMR doesn't work. So in this PR I add empty script to if HMR is enabled. Right now it doesn't matter if there is one already (not sure if I need to add checks for that).

Closes https://github.com/parcel-bundler/parcel/issues/943

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [ ] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

mischnic commented:
1. We should track whether there is already a script tag in the HTML file (`collectDependencies` could track that).
2. It shouldn't be an inline script tag, but rather this so that it becomes a separate bundle:
```js
ast.program.push({tag: 'script', attrs: {src: asset.addDependency("hmr-script"))}, content: []});
// ...
return [asset, {type: "js", uniqueKey: "hmr-script", content: ""}, ...];
```

faergeek commented:
Thanks for feedback!

1. Shouldn't then both `collectDependencies` and `extractInlineAssets` track if we have a script already? Because HMR code is injected into inline scripts too. And what's a good way to track that in your opinion? Pass an object with a boolean prop, which is mutated inside of those functions?

mischnic commented:
I would return a boolean in `collectDependencies` and return `{hasScript: true/false, assets: ...}` in `extractInlineAssets`

faergeek commented:
Now `hasScripts` flags are returned from `collectDependencies` and `extractInlineAssets` to ensure we need to add script tag for HMR to work. I also pushed an example to demonstrate invariant violation. If you run `yarn start` for it, the script is injected and everything is fine. But whenever you uncomment script tag in html, this invariant in `AssetGraph` is violated for inserted (and now being removed) node:
 
https://github.com/parcel-bundler/parcel/blob/e44a16888b8177af6d2863139a0096a0d4c5c72a/packages/core/core/src/AssetGraph.js#L162

I'm not sure if that's because I did something wrong. I'm new to parcel codebase. Looking for help

mischnic commented:
> this invariant in AssetGraph is violated for inserted

Your code is correct. I've run into this even without your PR, but haven't figured it out yet.

faergeek commented:
So the example is not needed anymore? Should I remove it? Please let me know if I need to do something more

mischnic commented:
Yes, remove that example. Instead add an integration test similar to this one: https://github.com/parcel-bundler/parcel/blob/90a3f012ccada0e5f80d9aac0dda3e08f21053d1/packages/core/integration-tests/test/html.js#L79-L94

In that assert call, the JS bundle should contain the `hmr-runtime.js` asset.

faergeek commented:
I removed an example and added integration test. There's also a react refresh runtime being injected. Is it correct? Also I'm not sure why pr benchmarks failed.

mischnic commented:
Add these two files:
```
packages/core/integration-tests/test/integration/html-no-js/yarn.lock
packages/core/integration-tests/test/integration/html-no-js/package.json
```
(yarn.lock is empty, package.json contains `{}`)

faergeek commented:
I added `package.json` and `yarn.lock`, and fixed test.

faergeek commented:
Is there anything else to be done with this PR?

mischnic commented:
No, but this apparently exaggerates the assertion error from https://github.com/parcel-bundler/parcel/pull/5414#issuecomment-742764000. So I'd like to get that fixed first before merging your PR.

faergeek commented:
@mischnic is there any issue to track this? I noticed that this problem only happens when assets returned from transform get removed. So the error happens if you remove inline script, but when you remove script with `src`, it doesn't. The same thing with `<link rel="stylesheet" href="...">` vs `<style>...</style>`. And when removing inline styles the error happens as well. So now I'm not sure why this invariant is needed there.

Also I noticed that `<style>...</style>` tags and inline styles don't get hot-reloaded. I'm seeing an error `Uncaught SyntaxError: Unexpected token '<'` in browser devtools.

mischnic commented:
PR to fix the assertion: https://github.com/parcel-bundler/parcel/pull/5522

I've opened a new issue with some findings about HMR update for inline assets: https://github.com/parcel-bundler/parcel/issues/5523

mischnic commented on packages/transformers/html/src/HTMLTransformer.js:
```diff
@@ -23,12 +23,45 @@ export default (new Transformer({
+      ast.program.push({
+        tag: 'script',
+        attrs: {
+          src: asset.addURLDependency('hmr.js', {
+            isAsync: false,
+            isEntry: false,
+            isIsolated: true,
+          }),
```
```suggestion
          src: asset.addURLDependency('hmr.js'),
```

mischnic commented on packages/core/integration-tests/test/html.js:
```diff
@@ -93,6 +93,28 @@ describe('html', function() {
+    let b = await bundle(
+      path.join(__dirname, '/integration/html-no-js/index.html'),
+      {
+        hot: {
+          port: 42,
+        },
```
```suggestion
        hot: {},
```

mischnic commented:
Thank you!

aarongeorge commented:
Hey everyone, thanks for the work on this.

There is an edge case that seems to be missing from this fix. If the file is a `.pug` (or anything other than `.html`) this HMR is still not injected.

I'm thinking we should be checking the entry file for a script tag, instead of the `.html` file for it (maybe we do both?)

Thoughts?

mischnic commented:
It should also work for Pug because the HTML that was generated from the Pug file get analyzed for these script tags.

faergeek commented:
Hey @aarongeorge 

I believe this is a better place to fix pug/jade https://github.com/parcel-bundler/parcel/issues/5523

faergeek commented:
@mischnic I'm not entirely sure, but I remember there were some checks for file extension. So I assume something may not work for pug now
